### PR TITLE
Add VQGAN-CLIP research project

### PR DESCRIPTION
**Pull request by ErwannMillon**
# What does this PR do?
Implements VQGAN-CLIP using huggingface CLIP models

Related to #21064 

This Research Project allows users to generate or edit images with a single line of code. It wraps the huggingface CLIPProcessor class, allowing images to be processed as torch tensors in order to preserve gradient flow through the transformations. 

Features:
- Positive and negative prompts
- Multiple prompts
- Prompt Weights
- Creating GIF animations of the transformations
- Wandb logging

Tagging @amyeroberts for review :)

### CODE REVIEW

File: examples/research_projects/vqgan-clip/loaders.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import importlib
+
+import torch
+
```
Comment (amyeroberts) [inline examples/research_projects/vqgan-clip/loaders.py:4]: nit 
```suggestion
```

File: examples/research_projects/vqgan-clip/img_processing.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    x = x.permute(1, 2, 0).numpy()
+    if process:
+        x = (255 * x).astype(np.uint8)
+    x = Image.fromarray(x)
+    if not x.mode == mode:
+        x = x.convert(mode)
+    return x
+
+
+def get_pil(x):
```
Comment (amyeroberts) [inline examples/research_projects/vqgan-clip/img_processing.py:42]: nit: function name is a bit misleading as it doesn't return a PIL image. The logic is also repeated in `custom_to_pil`.

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        - A list of prompts and weights, e.g [("A smiling woman", 1), ("a woman with brown hair", 3)]
+        """
+        if image_path:
+            self.latent = self._get_latent(image_path)
+        else:
+            self.latent = torch.randn(self.latent_dim, device=self.device)
+        if self.log:
+            self._init_logging(pos_prompts, neg_prompts, image_path)
+
+        assert pos_prompts, "You must provide at least one positive prompt."
```
Comment (amyeroberts) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:238]: This should be an exception as it's checking inputs - asserts are there for checking bugs in the code 🐛

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+    def _get_clip_similarity(self, prompts, image, weights=None):
+        clip_inputs = self.clip_preprocessor(text=prompts, images=image, return_tensors="pt", padding=True)
+        clip_outputs = self.clip(**clip_inputs)
+        similarity_logits = clip_outputs.logits_per_image
+        if weights is not None:
+            similarity_logits = similarity_logits * weights
+        return similarity_logits.sum()
+
+    def _get_CLIP_loss(self, pos_prompts, neg_prompts, image):
```
Comment (amyeroberts) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:147]: To be consistent with naming above

```suggestion
    def _get_clip_loss(self, pos_prompts, neg_prompts, image):
```

**Review by amyeroberts — COMMENTED** [review]
Nice - looking good! Thanks for adding this example. 

I've made a few comments, mostly nits. None are critical and I'll leave it to you to decide what you do or don't want to change. 

One thing to update is the `face.jpeg` image. To avoid adding lots of data to the repo, could you do the same as in other examples, and upload the image to a dataset where it can be display through hrefs and downloaded from e.g. https://github.com/huggingface/transformers/blob/42b60f8b02941b0c40c42e150a101eb372c3856e/examples/research_projects/codeparrot/README.md?plain=1#L3 ?

**Comment by ErwannMillon** [comment]
Hi there,
Thanks for the feedback, I fixed the style issues and removed the face image. 
Also refactored the code to have more accurate function names, used just the tokenizer, changed the assertions to exceptions, and removed some extraneous code (eg double crop, freeze_module)
Have a great day :)
Erwann

**Comment by amyeroberts** [comment]
@ErwannMillon - thanks for the updates. It's looking good 😎 ! 

Just two last things before I think we're ready to merge: 
* Removing the `face.jpg` file and instead pointing to a place where it can be downloaded
* Could you also remove the notebook? Like the image, we want to avoid adding large files as much as possible. Happy for you to link to e.g. a colab which shows this demo in the README.

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+class ProcessorGradientFlow:
+    """
+    This wraps the huggingface CLIP processor to allow backprop through the image processing step.
+    The original processor forces conversion to PIL images, which is faster for image processing but breaks gradient flow.
+    We call the original processor to get the text embeddings, but use our own image processing to keep images as torch tensors.
+    """
+
+    def __init__(self, device: str = "cpu", clip_model: str = "openai/clip-vit-large-patch14") -> None:
+        self.device = device
+        self.tokenizer = CLIPTokenizerFast.from_pretrained(clip_model)
```
Comment (amyeroberts) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:26]: If using the fast tokenizers, you'll need to add the `tokenizers` library as a requirement

Comment (ErwannMillon) [reply examples/research_projects/vqgan-clip/VQGAN_CLIP.py:26]: Done

**Review by amyeroberts — APPROVED** [review]
Looks good to merge from my end :)

**Review by sgugger — APPROVED** [review]
Thanks for working on this! Can we just remove the empty folder `model_checkpoints` from the PR?

**Review by sgugger — APPROVED** [review]
